win32/derive/src/trace.rs

```diff
@@ -1,34 +1,75 @@
-use quote::quote;
+use quote::{format_ident, quote};
 
 /// Insert a call to trace::trace() at the front of a function that logs its name and arguments.
 // TODO: this maybe belongs at the "impls" mod, so it can do its own traversal of the stack (?).
-pub fn add_trace(func: &mut syn::ItemFn) {
-    let name = func.sig.ident.to_string();
-    let mut args: Vec<&syn::Ident> = Vec::new();
-    for arg in func.sig.inputs.iter().skip(1) {
+pub fn add_trace(mut func: syn::ItemFn) -> proc_macro2::TokenStream {
+    let name = &func.sig.ident;
+    let name_string = name.to_string();
+    let mut arg_names: Vec<&syn::Ident> = Vec::new();
+    for arg in &func.sig.inputs {
         match arg {
             syn::FnArg::Typed(arg) => match &*arg.pat {
                 syn::Pat::Ident(pat) => {
-                    args.push(&pat.ident);
+                    arg_names.push(&pat.ident);
                 }
                 _ => {}
             },
             _ => {}
         };
     }
-    let synargs = args
+    let trace_arg_values = arg_names
         .iter()
+        .skip(1)
         .map(|arg| {
             let name = arg.to_string();
             quote!((#name, &#arg))
         })
         .collect::<Vec<_>>();
-    let arg_count = args.len();
-    let stmt: syn::Stmt = syn::parse_quote! {
-        if crate::trace::enabled(TRACE_CONTEXT) {
-            let args: &[(&str, &dyn std::fmt::Debug); #arg_count] = &[#(#synargs),*];
-            crate::trace::trace(TRACE_CONTEXT, std::file!(), std::line!(), #name, args);
+    let trace_arg_count = trace_arg_values.len();
+    let prolog = quote! {
+        let __trace_context = if crate::trace::enabled(TRACE_CONTEXT) {
+            let args: &[(&str, &dyn std::fmt::Debug); #trace_arg_count] = &[#(#trace_arg_values),*];
+            Some(crate::trace::trace_begin(TRACE_CONTEXT, #name_string, args))
+        } else {
+            None
+        };
+    };
+    let epilog = quote! {
+        if let Some(__trace_context) = __trace_context {
+            crate::trace::trace_return(&__trace_context, std::file!(), std::line!(), &__ret);
         }
     };
-    func.block.stmts.insert(0, stmt);
+
+    let trace_vis = func.vis.clone();
+    let mut trace_sig = func.sig.clone();
+    // Remove mutability and references from arguments.
+    for arg in &mut trace_sig.inputs {
+        match arg {
+            syn::FnArg::Typed(arg) => match &mut *arg.pat {
+                syn::Pat::Ident(pat) => {
+                    pat.by_ref = None;
+                    pat.mutability = None;
+                }
+                _ => {}
+            },
+            _ => {}
+        };
+    }
+
+    let impl_name = format_ident!("{}_impl", name);
```

Comment:
I think you can avoid needing to generate a name here by generating code like

```
#prolog
let ret = (|arg1, arg2| { #func })(arg1, arg);
#epilog
```

I think what you have is fine too, just suggesting it in case you like it more

Reply:
I did it this way so that inner `return`s don’t return the outer fn. I don’t remember if closures would work the same, tbh